### DISCUSSION TITLE
How to map molecule fragments back to parent molecule?

### DISCUSSION THREAD

**Discussion by colinrsmall** [rdkit/rdkit#5458] [Q&A] [answered]
I have a molecule from which I am generating fragments/r-groups, and I'd like to retrieve the atom numbers of the groups/fragments from the parent molecule. I've tried two ways of generating the fragments:

rdRGroupDecomposition.RGroupDecompose:
```
mol = Chem.MolFromSmiles("CC(=O)N1CCC(Nc2cncc(C(=O)c3cn(C(C)C)c4ncnc(N)c34)n2)C1c1ccc(F)cc1")
core = MurckoScaffold.GetScaffoldForMol(mol)
groups = rdRGroupDecomposition.RGroupDecompose([core], [mol], asRows=False)
```

Chem.ReplaceCore + Chem.GetMolFrags:
```
mol = Chem.MolFromSmiles("CC(=O)N1CCC(Nc2cncc(C(=O)c3cn(C(C)C)c4ncnc(N)c34)n2)C1c1ccc(F)cc1")
core = MurckoScaffold.GetScaffoldForMol(mol)
floating_groups = Chem.ReplaceCore(mol,core,labelByIndex=True)
groups = Chem.GetMolFrags(chains,asMols=True)
```

However, using the groups/fragments generated by either method with GetSubstructMatch fails to produce any matches:

rdRGroupDecomposition.RGroupDecompose:
```
fragment = groups[0]["R1"][0]
mol.GetSubstructMatch(fragment)
> ()
```

Chem.ReplaceCore + Chem.GetMolFrags:
```
fragment = groups[0]
mol.GetSubstructMatch(fragment)
> ()
```

I'm a little new to computational chemistry. Does anyone know how to properly map the fragments generated by either method (preferably those generated with rdRGroupDecomposition.RGroupDecompose) back to the parent molecule and get the atom numbers of the fragment?

**Comment by ptosco** [comment, marked as answer]
@colinrsmall Using `GetSubStructMatch()` has a high chance to retrieve indices unrelated to your R groups if you have symmetries in your molecule; additionally, the presence of explicit Hs and dummy atoms further complicates things.
You are better off labeling atoms with their original indices ahead of doing the R-group decomposition, and use those to do the mapping afterwards, e.g.:
```

from rdkit import Chem
from rdkit.Chem.Scaffolds import MurckoScaffold
from rdkit.Chem import rdRGroupDecomposition

mol = Chem.MolFromSmiles("CC(=O)N1CCC(Nc2cncc(C(=O)c3cn(C(C)C)c4ncnc(N)c34)n2)C1c1ccc(F)cc1")

for a in mol.GetAtoms():
    a.SetIntProp("__origIdx", a.GetIdx())

core = MurckoScaffold.GetScaffoldForMol(mol)

groups = rdRGroupDecomposition.RGroupDecompose([core], [mol], asRows=False)

for rlabel, rgroups in filter(lambda x: x[0].startswith("R"), groups[0].items()):
    for rgroup in rgroups:
        print(rlabel, [(a.GetIdx(), a.GetIntProp("__origIdx")) for a in rgroup.GetAtoms() if a.HasProp("__origIdx")])
R1 [(0, 34)]
R2 [(0, 26)]
R3 [(0, 18), (1, 19), (2, 20)]
R4 [(0, 0), (1, 1), (2, 2)]
```

**Reply by colinrsmall** [reply]
Many thanks. I didn't know that atom properties would be carried over after breaking the mol into fragments.

Just to better my understanding of rdkit, do you know why the fragments as generated by either approach aren't found as substruct matches of the original molecule?

**Reply by ptosco** [reply]
You only need to turn dummy atoms into query atoms such that they actually match any atom:
```
mols = [mol]

# this fails, as you have found out
for rlabel, rgroups in filter(lambda x: x[0].startswith("R"), groups[0].items()):
    print(rlabel, [mols[i].HasSubstructMatch(rgroup) for i, rgroup in enumerate(rgroups)])
R1 [False]
R2 [False]
R3 [False]
R4 [False]

params = Chem.AdjustQueryParameters.NoAdjustments()
params.makeDummiesQueries = True
adjusted_rgroups = {rlabel: [Chem.AdjustQueryProperties(rgroup, params)
    for rgroup in rgroups] for rlabel, rgroups in filter(lambda x: x[0].startswith("R"), groups[0].items())}

# This will now work as expected
for rlabel, rgroups in adjusted_rgroups.items():
    print(rlabel, [mols[i].HasSubstructMatch(rgroup) for i, rgroup in enumerate(rgroups)])
R1 [True]
R2 [True]
R3 [True]
R4 [True]
```
That said, you will still have the symmetry problem, and fragments such as `[H]-*` will actually fail to match since the starting molecule does not have explicit hydrogens. In general, as I already mentioned, I would recommend against such an approach as it may lead to multiple problems.
